main.tf

```diff
@@ -0,0 +1,314 @@
+  provisioner "local-exec" {
+    when    = "destroy"
+    command = "sh ${path.module}/scripts/delete-bq-tables.sh ${local.destination_project} ${local.destination_name}"
+  }
+
+  depends_on = [
+    "null_resource.enable_destination_api",
+  ]
+}
+
+# Storage bucket
+resource "google_storage_bucket" "bucket" {
+  count         = "${local.destination_type == "storage" ? 1 : 0}"
+  name          = "${local.destination_name}"
+  project       = "${local.destination_project}"
+  storage_class = "MULTI_REGIONAL"
+  location      = "US"
+  force_destroy = true
+
+  depends_on = [
+    "null_resource.enable_destination_api",
+  ]
+}
+
+#---------#
+# Logsink #
+#---------#
+
+# Project-level
+resource "google_logging_project_sink" "sink" {
+  count                  = "${local.is_project_level ? 1 : 0}"
+  name                   = "${var.name}"
+  project                = "${var.project}"
+  filter                 = "${var.filter}"
+  destination            = "${local.destination_uri}"
+  unique_writer_identity = "${var.unique_writer_identity}"
+
+  depends_on = [
+    "google_storage_bucket.bucket",
+    "google_pubsub_topic.topic",
+    "google_bigquery_dataset.dataset",
+  ]
+}
+
+# Folder-level
+resource "google_logging_folder_sink" "sink" {
+  count            = "${local.is_folder_level ? 1 : 0}"
+  name             = "${var.name}"
+  folder           = "${var.folder}"
+  filter           = "${var.filter}"
+  include_children = "${var.include_children}"
+  destination      = "${local.destination_uri}"
+
+  depends_on = [
+    "google_storage_bucket.bucket",
+    "google_pubsub_topic.topic",
+    "google_bigquery_dataset.dataset",
+  ]
+}
+
+# Org-level
+resource "google_logging_organization_sink" "sink" {
+  count            = "${local.is_org_level ? 1 : 0}"
+  name             = "${var.name}"
+  org_id           = "${var.org_id}"
+  filter           = "${var.filter}"
+  include_children = "${var.include_children}"
+  destination      = "${local.destination_uri}"
+
+  depends_on = [
+    "google_storage_bucket.bucket",
+    "google_pubsub_topic.topic",
+    "google_bigquery_dataset.dataset",
+  ]
+}
+
+# Billing-level
+# resource "google_logging_billing_account_sink" "sink" {
+#   count           = "${local.is_billing_level ? 1 : 0}"
+#   name            = "${var.name}"
+#   billing_account = "${var.billing_id}"
+#   filter          = "${var.filter}"
+#   destination     = "${local.destination_uri}"
+#
+#   depends_on = [
+#     "google_storage_bucket.bucket",
+#     "google_pubsub_topic.topic",
+#     "google_bigquery_dataset.dataset",
+#   ]
+# }
+
+#------------------------------#
+# Service account IAM bindings #
+#------------------------------#
+
+# Org-level
+resource "google_project_iam_member" "sink_member_org" {
+  count   = "${local.is_org_level ? 1 : 0}"
+  project = "${local.destination_project}"
+  role    = "${local.role}"
+  member  = "${google_logging_organization_sink.sink.writer_identity}"
```

Comment:
Looks like `member` is the only field which differs between these. I bet we could combine all 3 iam_member resources into one.

Reply:
Problem here is that the only one of the 3 resources will be defined. I encountered an error trying to reference resources that are not created (because count is 0).

Reply:
We've worked around this on another modules using splat syntax: https://www.terraform.io/docs/configuration/interpolation.html#attributes-of-other-resources

Not a blocker, but just a suggestion to give it a try.

Reply:
Nevermind, I actually solved that already (look at sink_output['writer']) variable in the code, that's exactly the 'member' that we need :)

Reply:
The 3 resources become:

```
resource "google_project_iam_member" "sink_member" {
  project = "${local.destination_project}"
  role    = "${local.role}"
  member  = "${local.sink_output["writer"]}"
}
```

Reply:
Looking more at it, I think we should use `google_storage_bucket_iam_member` and equivalents for pubsub and bigquery to restrict permissions here. So still 3 resources needed.

Reply:
Filed [a bug](https://github.com/terraform-providers/terraform-provider-google/issues/2051) for `bigquery_dataset_iam_member` and `bigquery_dataset_iam_binding` to be added.

Reply:
Ah, I see that the resource types will differ between the different types so we'll stick with 3 resources.